### DISCUSSION TITLE
Help with integrating Stockfish with android application (Kotlin or Java)

### DISCUSSION THREAD

**Discussion by seekkii** [official-stockfish/Stockfish#4605] [Q&A]
Hello, I have created a share library file named libstockfish.so with all the executables downloaded from the official website by syncing  Cmakelist,txt with graddle. I put the shared library inside appropriate architecture folders, arm64-v8a, armeabi-v7a, x86, x86_64, the compiler was able to recognize the file and I managed to load it without any exception:
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
val file = File( applicationContext.applicationInfo.nativeLibraryDir + "/libstockfish.so")
var process: Process? = null
try {
        process = Runtime.getRuntime().exec(file.path)
} catch (e: IOException) {
            e.printStackTrace()
}
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Then I make a thread that handle command lines which will be passed and write it to line:
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 val outThread = Thread(Runnable {
            val processOut = process ?: return@Runnable
            val scanner = Scanner(processOut.inputStream)
            try {
                while (scanner.hasNextLine()) {
                    val line = scanner.nextLine()
                    Log.d("MainActivity", "Received output: $line")
                }
            } catch (e: IOException) {
                Log.e("MainActivity", "Error reading output", e)
            } finally {
                scanner.close()
            }
        })
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Then I start the thread and pass the command in by writing to the thread outputstream, but the received result from process.inputstream is null.
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 outThread.start()
 var command = "uci\n"
        try{
            if (process != null) {
                process.outputStream.write(command.toByteArray())
                process.outputStream.flush()
            }
        }
        catch (e: IOException) { }
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
So I would like to ask if this behavior is expected? ,or am I doing something wrong? , any help will be greatly appreciated!

**Comment by Disservin** [comment]
Eh i dont know any java but looking at this makes me wonder why you write to outputstream ? You have to write to stdin and stockfish's output will be stdout. There's some small documentation on how to interact with Stockfish from your own app.
 https://github.com/official-stockfish/Stockfish/wiki/Stockfish-FAQ#how-can-i-use-stockfish-in-my-own-software-application

**Reply by seekkii** [reply]
Hello, thank you a lot for the reply!
Process.outputstream is actually the outputstream connected to the input of the subprocess, so writing to it should be similar to writing to stdin of the process, if I'm not mistaken, and reading from the process.inputstream should be similar to reading from stdout of the process.

**Comment by dav1312** [comment]
Since I don't think there are many Android developers here, you might want to take a look at how other developers did it
- Droidfish app: https://github.com/peterosterlund2/droidfish
- Java examples: https://github.com/search?q=stockfish+lang%3AJava&type=code
- Kotlin examples: https://github.com/search?q=stockfish+lang%3AKotlin+&type=code

**Comment by wikbpjwstk** [comment]
hello, haw you create libstockfish.so. I have problem with it. can you help?

**Reply by seekkii** [reply]
Hello, depends on what you are using, but there are 2 options for me 
Download stockfish
Navigate to the folder contains src and run make as how they instruct you( Keep in mind that they have different options for different platforms)
After make is finished you should see a bunch of .o
run g++ libstockfish.so evaluate.o ....... all the .o here

If you use android studio:
+Create native c++ project, specify c++17 if you use the latest stockfish source code
+Put Stockfish src inside cpp folder
+In Cmakelist.txt 
         add_library{stockfish evaluate.cpp evaluate.h..... all the excutable here, don't include any nnue related file if you don't need it}
         If you need nnue however add this under add_library:
         add_custom_target(
             copy_nnue_file ALL
             DEPENDS ${CMAKE_CURRENT_BINARY_DIR}/nn-ad9b42354671.nnue
          )
          add_dependencies(stockfish copy_nnue_file)
+In graddle.build
        Add this to your android{  
        sourceSets {
        main {
            jniLibs.srcDirs = ['libs']
        }
    }
      
       externalNativeBuild {
        cmake {
            // Specifies the path to your CMake build script (CMakeLists.txt).
            path "src/main/cpp/CMakeLists.txt"
        }
    }
    And this to your dependency{
           implementation fileTree(dir: "libs", include: ["*.jar", "*.so"])
+In manifest 
<application
        android:extractNativeLibs="true"
        ............
</application>
 Build->Build apk->locate, change apk to zip, open it and you should see a lib folder that contains 4 folders, libstockfish.so for each of the architecture is inside.

**Reply by wikbpjwstk** [reply]
I tried to do everything as in the manual. the build process compiles but no *.so files are generated. someone can help more.I tried to do everything as in the manual. the build process compiles but no *.so files are generated. someone can help more.


**build.gradle**
plugins {
    id 'com.android.application'
}

android {
    namespace 'com.example.stockfish'
    compileSdk 33

    defaultConfig {
        applicationId "com.example.stockfish"
        minSdk 29
        targetSdk 33
        versionCode 1
        versionName "1.0"

        testInstrumentationRunner "androidx.test.runner.AndroidJUnitRunner"
        externalNativeBuild {
            cmake {

                cppFlags '-std=c++17'

            }
        }

        sourceSets {
            main {
                jniLibs.srcDirs = ['libs']
            }
        }
    }

    buildTypes {
        release {
            minifyEnabled false
            proguardFiles getDefaultProguardFile('proguard-android-optimize.txt'), 'proguard-rules.pro'
        }
    }
    compileOptions {
        sourceCompatibility JavaVersion.VERSION_1_8
        targetCompatibility JavaVersion.VERSION_1_8
    }
    externalNativeBuild {
        cmake {
            path file('src/main/cpp/CMakeLists.txt')

            version '3.22.1'
        }
    }
    buildFeatures {
        viewBinding true
    }
}

dependencies {

    implementation 'androidx.appcompat:appcompat:1.6.1'
    implementation 'com.google.android.material:material:1.9.0'
    implementation fileTree(dir: "libs", include: ["*.jar", "*.so"])
    implementation 'androidx.constraintlayout:constraintlayout:2.1.4'
    testImplementation 'junit:junit:4.13.2'
    androidTestImplementation 'androidx.test.ext:junit:1.1.5'
    androidTestImplementation 'androidx.test.espresso:espresso-core:3.5.1'
}

**CMAKE**
# For more information about using CMake with Android Studio, read the
# documentation: https://d.android.com/studio/projects/add-native-code.html

# Sets the minimum version of CMake required to build the native library.

cmake_minimum_required(VERSION 3.22.1)

# Declares and names the project.

project("stockfish")

# Creates and names a library, sets it as either STATIC
# or SHARED, and provides the relative paths to its source code.
# You can define multiple libraries, and CMake builds them for you.
# Gradle automatically packages shared libraries with your APK.

add_library( # Sets the name of the library.
        stockfish

        # Sets the library as a shared library.
#                SHARED

        # Provides a relative path to your source file(s).
        benchmark.cpp
        bitbase.cpp
        bitboard.cpp
        bitboard.h
        CMakeLists.txt
        endgame.cpp
        endgame.h
        evaluate.cpp
        evaluate.h
        main.cpp
        material.cpp
        material.h
        misc.cpp
        misc.h
        movegen.cpp
        movegen.h
        movepick.cpp
        movepick.h
        pawns.cpp
        pawns.h
        position.cpp
        position.h
        psqt.cpp
        psqt.h
        search.cpp
        search.h
        thread.cpp
        thread.h
        thread_win32_osx.h
        timeman.cpp
        timeman.h
        tt.cpp
        tt.h
        tune.cpp
        tune.h
        types.h
        uci.cpp
        uci.h
        ucioption.cpp)

add_custom_target(
        copy_nnue_file ALL
        DEPENDS ${CMAKE_CURRENT_BINARY_DIR}/nn-ad9b42354671.nnue
)
add_dependencies(stockfish copy_nnue_file)


# Searches for a specified prebuilt library and stores the path as a
# variable. Because CMake includes system libraries in the search path by
# default, you only need to specify the name of the public NDK library
# you want to add. CMake verifies that the library exists before
# completing its build.

find_library( # Sets the name of the path variable.
        log-lib

        # Specifies the name of the NDK library that
        # you want CMake to locate.
        log)

# Specifies libraries CMake should link to your target library. You
# can link multiple libraries, such as libraries you define in this
# build script, prebuilt third-party libraries, or system libraries.

target_link_libraries( # Specifies the target library.
        stockfish

        # Links the target library to the log library
        # included in the NDK.
        ${log-lib})

**Reply by wikbpjwstk** [reply]
i have this issue

ninja: error: 'nn-ad9b42354671.nnue', needed by 'CMakeFiles/copy_nnue_file', missing and no known rule to make it

**Reply by Jochengehtab** [reply]
Hello @seekkii can you pleas describe more Detail how i can use the nnue when i am compiling it with CMake. Thanks a lot

**Comment by wikbpjwstk** [comment]
and after :
add_custom_target(
        copy_nnue_file ALL
        DEPENDS ${PROJECT_SOURCE_DIR}/incbin/nn-ad9b42354671.nnue
)
add_dependencies(stockfish copy_nnue_file)

**Comment by wikbpjwstk** [comment]
and after :::::

add_custom_target(
        copy_nnue_file ALL
        DEPENDS ${PROJECT_SOURCE_DIR}/incbin/nn-ad9b42354671.nnue
)
add_dependencies(stockfish copy_nnue_file)


C/C++: <inline asm>:6:9: error: Could not find incbin file 'nn-ad9b42354671.nnue'
C/C++: .incbin "nn-ad9b42354671.nnue"
C/C++:         ^
C/C++: 1 error generated.

**Reply by seekkii** [reply]
you should copy the nn-ad9b42354671.nnue into cbin folder before running those if you want to use nnue. There should be a lib folder when you inspect your apk, but if not go to app/.cxx/debug/something/, inside there are 4 folders (arm64v8a, armaebiv7a, x86, x86_64, in each of them find stockfish.dir inside cmakefile, change their name to libstockfish.so, before copying those 4 files(arm64v8a....) into your libs folder in app/libs.

**Reply by wikbpjwstk** [reply]
This not working. I have this stockfish.dir directory but i need file with extension .so . What to do?

**Reply by seekkii** [reply]
rename it to libstockfish.so add dependency dependencies {other dependencies...... implementation fileTree(dir: 'libs', include: '*.so')

**Comment by wikbpjwstk** [comment]
@seekkii how to do it without this nnue file?

**Reply by seekkii** [reply]
I didn't try using graddle to exclude nnue file but i assume you should remove every nnue related file inside add_library{} tag, if there is nnue related stuff inside some part of the src code you might need to comment them as well. 

Alternatively just do something like g++ libstockfish.so evaluate.o blah.o blah.o, remember to exclude the nnue related files though

**Reply by wikbpjwstk** [reply]
I close my  question and i  solved problems
